kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java

```diff
@@ -850,22 +861,94 @@ public Term parse(String input, Source source, int startLine, int startColumn) {
       // a parse error after the outer loop finishes also.
       if (S.get(k+1).empty() && Q.empty()) {
         // no more states to process, so it must be a parse error
-        parseError(data, k);
+        parseError(data, S, k);
       }
     }
     if (S.get(data.words.length).empty()) {
-      parseError(data, data.words.length-1);
+      parseError(data, S, data.words.length-1);
     }
     // finished parsing successfully, so return the final parse forest
     return Ambiguity.apply(S.get(data.words.length).states.get(0).parseTree().stream().map(list -> list.get(0)).collect(Collectors.toSet()));
   }
 
+  // We are only interested in displaying states that span the entire input
+  // when a parse error occurs; such states have a start-index of 0.
+  private Set<EarleyState> spanningStates(EarleySet parses) {
+    return parses.states.stream()
+            .filter(state -> state.start == 0)
+            .collect(Collectors.toSet());
+  }
+
+  // We heuristically identify the best state-set for producing diagnostics as the
+  // most recent such set that includes a _spanning state_; i.e. one with a start
+  // index of zero.
+  private Set<EarleyState> bestDiagnosticStates(List<EarleySet> S, int k) {
+    for(int i = k; i >= 0; --i) {
+      Set<EarleyState> candidate = spanningStates(S.get(i));
+      if(!candidate.isEmpty()) {
+        return candidate;
+      }
+    }
+
+    return new HashSet<>();
+  }
+
+  private String partialParseTreesDiagnostic(Set<EarleyState> spanningStates) {
+    if(spanningStates.isEmpty()) {
+      return "No top-level production could apply to this input.";
+    }
+
+    StringBuilder msg = new StringBuilder();
+
+    // In the case of a parse error, we have a partial parse tree constructed
+    // that comprises each of the successful non-terminals in the current rule.
+    // To print an AST with correct arity, we want to create a dummy term that
+    // will never type-check, but can sit in a parsed AST to represent the errored
+    // non-terminal, as well as the ones we have not yet attempted to parse.
+    Production errorProd = Production.apply(Seq(), Sort("<invalid>"), Seq(), Att());
+    Term error = Constant.apply("<error>", errorProd);
+    Term incomplete = Constant.apply("<incomplete>", errorProd);
+
+    for(var state : spanningStates) {
+      msg.append("  Attempting to apply production:\n    ").append(state.prod).append("\n");
+      for(var possibleTree : state.parseTree()) {
```

Comment:
I'm not sure I like what is going on here. I understand you did this in order to try to make the term have the correct arity, but I'm not sure this is sufficient in order to do so. I see you adding children to exactly one top-level application, but there's no particular reason to assume that all terms that are missing children because they haven't parsed yet are at the top level. So I'm not sure this change is sufficient, but I'm also not sure that it's needed. We're just printing the abstract syntax so it shouldn't need to have the correct arity, right?